xtm/xtm_api.cc

```diff
@@ -0,0 +1,380 @@
+};
+
+int
+xtm_queue_delete(struct xtm_queue *queue)
+{
+	int rc = 0;
+	if (((queue->flags & XTM_QUEUE_NEED_TO_CLOSE_READFD) != 0) &&
+	    ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) != 0) &&
+	    close(queue->producer_read_fd) < 0)
+		rc = -1;
+	if (((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) != 0) &&
+	    queue->producer_write_fd != queue->producer_read_fd &&
+	    close(queue->producer_write_fd) < 0)
+		rc = -1;
+	if (((queue->flags & XTM_QUEUE_NEED_TO_CLOSE_READFD) != 0) &&
+	    close(queue->consumer_read_fd) < 0)
+		rc = -1;
+	if (queue->consumer_read_fd != queue->consumer_write_fd &&
+	    close(queue->consumer_write_fd) < 0)
+		rc = -1;
+	free(queue);
+	return rc;
+}
+
+int
+xtm_queue_notify_consumer(struct xtm_queue *queue)
+{
+	return notify_fd(queue->consumer_write_fd);
+}
+
+int
+xtm_queue_notify_producer(struct xtm_queue *queue)
+{
+	return notify_fd(queue->producer_write_fd);
+}
+
+int
+xtm_queue_probe(struct xtm_queue *queue)
+{
+	if (queue->queue.free_count() == 0) {
+		errno = ENOBUFS;
+		return -1;
+	}
+	return 0;
+}
+
+unsigned
+xtm_queue_count(struct xtm_queue *queue)
+{
+	return queue->queue.count();
+}
+
+int
+xtm_queue_push_fun(struct xtm_queue *queue, xtm_queue_fun_t fun,
+		   void *fun_arg)
+{
+	union xtm_msg xtm_msg;
+	xtm_msg.fun = fun;
+	xtm_msg.fun_arg = fun_arg;
+	if (queue->queue.put(&xtm_msg, 1) == 1)
+		return 0;
+	errno = ENOBUFS;
+	if ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) != 0) {
+		__atomic_store_n(&queue->is_notification_expected,
+				 true, __ATOMIC_RELAXED);
+	}
+	__atomic_thread_fence(__ATOMIC_SEQ_CST);
+	return (queue->queue.put(&xtm_msg, 1) == 1 ? 0 : -1);
+}
+
+int
+xtm_queue_consumer_fd(struct xtm_queue *queue)
+{
+	return queue->consumer_read_fd;
+}
+
+int
+xtm_queue_producer_fd(struct xtm_queue *queue)
+{
+	return queue->producer_read_fd;
+}
+
+int
+xtm_queue_invoke_funs_all(struct xtm_queue *queue)
+{
+	struct xtm_scsp_queue_read_iterator<xtm_msg> iter;
+	const union xtm_msg *xtm_msg;
+	unsigned cnt = 0;
+
+	iter.begin(&queue->queue);
+	while((xtm_msg = iter.read()) != nullptr) {
+		xtm_msg->fun(xtm_msg->fun_arg);
+		cnt++;
+	}
+	iter.end();
+	/**
+	 * In case queue was full, we need full memory barrier,
+	 * to avoid a situation where due to reordering the code
+	 * we read queue->is_notification_expected, before storing
+	 * iterator last position in iter.end(). In case queue was
+	 * not full, we can not be afraid of the situation, when
+	 * kernel switch this thread here, because context switching
+	 * includes full memory barrier.
+	 */
+	if (iter.is_queue_was_full())
+		__atomic_thread_fence(__ATOMIC_SEQ_CST);
+	if ((queue->flags & XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS) == 0 ||
+	    cnt == 0 ||
+	    !__atomic_load_n(&queue->is_notification_expected,
+			      __ATOMIC_RELAXED))
+		goto finish;
+
+	__atomic_store_n(&queue->is_notification_expected, false,
+			 __ATOMIC_RELAXED);
+	if (notify_fd(queue->producer_write_fd) < 0)
+		return -(cnt + 1);
```

Comment:
* Before this all you could check if XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS is set. If it is not, you can finish right away and not do any fences, stores, etc;
* When you do still do anything if cnt == 0? Shouldn't the absence of XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS completely disable these notification? In the doc you said:
```
/**
 * Flag indicates, that producer thread want to receive
 * notifications when queue is not full.
 */
```
And yet you notify if count was 0. What is the purpose of this flag then?
* When you do an early check for XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS, you can delete the explicit fence and move SEQ_CST into the load operation. You could also turn it into compare-exchange instead of separate load + if + store + if. Would work?

All the same for pop_ptrs.

Reply:
Fixed